core/src/main/java/com/linecorp/armeria/internal/common/AbstractSplitHttpMessage.java

```diff
@@ -0,0 +1,343 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import static com.linecorp.armeria.common.util.Exceptions.throwIfFatal;
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpMessage;
+import com.linecorp.armeria.common.HttpObject;
+import com.linecorp.armeria.common.SplitHttpMessage;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.stream.NoopSubscriber;
+import com.linecorp.armeria.common.stream.StreamMessage;
+import com.linecorp.armeria.common.stream.SubscriptionOption;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+import com.linecorp.armeria.internal.common.stream.NoopSubscription;
+import com.linecorp.armeria.unsafe.PooledObjects;
+
+import io.netty.util.concurrent.EventExecutor;
+
+public abstract class AbstractSplitHttpMessage implements SplitHttpMessage, StreamMessage<HttpData> {
+
+    private static final Logger logger = LoggerFactory.getLogger(AbstractSplitHttpMessage.class);
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<BodySubscriber, Subscriber>
+            downstreamUpdater = AtomicReferenceFieldUpdater.newUpdater(BodySubscriber.class, Subscriber.class,
+                                                                       "downstream");
+
+    @SuppressWarnings("rawtypes")
+    protected static final AtomicReferenceFieldUpdater<AbstractSplitHttpMessage, HeadersFuture>
+            trailersFutureUpdater = AtomicReferenceFieldUpdater.newUpdater(AbstractSplitHttpMessage.class,
+                                                                           HeadersFuture.class,
+                                                                           "trailersFuture");
+
+    private final HttpMessage message;
+    protected final EventExecutor upstreamExecutor;
+
+    @Nullable
+    protected volatile HeadersFuture<HttpHeaders> trailersFuture;
+
+    private volatile boolean wroteAny;
+
+    protected AbstractSplitHttpMessage(HttpMessage message, EventExecutor executor) {
+        this.message = requireNonNull(message, "message");
```

Comment:
`message` is a bit vague to me. How about using `upstream`?

Reply:
It's good idea! thank you :)